internal/tls/config.go

```diff
@@ -0,0 +1,137 @@
+// - caFile, for the CA certificate
+//
+// Secrets with no certificate, private key, AND CA cert are ignored. If only a
+// certificate OR private key is found, an error is returned.
+func TLSClientConfigFromSecret(secret corev1.Secret, url string) (*tls.Config, *TLSBytes, error) {
+	return tlsClientConfigFromSecret(secret, url, false)
+}
+
+// tlsClientConfigFromSecret attempts to construct a TLS client config
+// for the given Secret. It returns the TLS client config or an error.
+//
+// kubernetesTLSKeys is a boolean indicating whether to check the Secret
+// for keys expected to be present in a Kubernetes TLS Secret. Based on its
+// value, the Secret is checked for the following keys:
+// - tls.key/keyFile for the private key
+// - tls.crt/certFile for the certificate
+// - ca.crt/caFile for the CA certificate
+// The keys should adhere to a single convention, i.e. a Secret with tls.key
+// and certFile is invalid.
+func tlsClientConfigFromSecret(secret corev1.Secret, url string, kubernetesTLSKeys bool) (*tls.Config, *TLSBytes, error) {
+	// Only Secrets of type Opaque and TLS are allowed. We also allow Secrets with a blank
+	// type, to avoid having to specify the type of the Secret for every test case.
+	// Since a real Kubernetes Secret is of type Opaque by default, its safe to allow this.
+	switch secret.Type {
+	case corev1.SecretTypeOpaque, corev1.SecretTypeTLS, "":
+	default:
+		return nil, nil, fmt.Errorf("cannot use secret '%s' to construct TLS config: invalid secret type: '%s'", secret.Name, secret.Type)
+	}
+
+	var certBytes, keyBytes, caBytes []byte
+	if kubernetesTLSKeys {
+		certBytes, keyBytes, caBytes = secret.Data[corev1.TLSCertKey], secret.Data[corev1.TLSPrivateKeyKey], secret.Data[CACrtKey]
+	} else {
+		certBytes, keyBytes, caBytes = secret.Data["certFile"], secret.Data["keyFile"], secret.Data["caFile"]
+	}
+
+	switch {
+	case len(certBytes)+len(keyBytes)+len(caBytes) == 0:
+		return nil, nil, nil
+	case (len(certBytes) > 0 && len(keyBytes) == 0) || (len(keyBytes) > 0 && len(certBytes) == 0):
+		return nil, nil, fmt.Errorf("invalid '%s' secret data: both certificate and private key need to be provided",
+			secret.Name)
+	}
+
+	tlsConf := &tls.Config{}
+	if len(certBytes) > 0 && len(keyBytes) > 0 {
+		cert, err := tls.X509KeyPair(certBytes, keyBytes)
+		if err != nil {
+			return nil, nil, err
+		}
+		tlsConf.Certificates = append(tlsConf.Certificates, cert)
+	}
+
+	if len(caBytes) > 0 {
+		cp, err := x509.SystemCertPool()
+		if err != nil {
+			return nil, nil, fmt.Errorf("cannot retrieve system certificate pool: %w", err)
+		}
+		if !cp.AppendCertsFromPEM(caBytes) {
+			return nil, nil, fmt.Errorf("cannot append certificate into certificate pool: invalid CA certificate")
+		}
+
+		tlsConf.RootCAs = cp
+	}
+
+	if url != "" {
+		u, err := neturl.Parse(url)
+		if err != nil {
+			return nil, nil, fmt.Errorf("cannot parse repository URL: %w", err)
+		}
+
+		tlsConf.ServerName = u.Hostname()
+	}
```

Comment:
i'm not sure we need it. i included it because the code that this is copied from had it. but, i haven't seen this particular logic in any other TLS client configuration code across the project:
* https://github.com/fluxcd/image-reflector-controller/blob/e2f5bb526a9f73716fe6028f645a514590a9f2cb/internal/secret/secret.go#L51
* https://github.com/fluxcd/pkg/blob/a4b7a305e47d167f82dd6702a154fd35d5aef5c6/runtime/tls/tls.go#L37

but yes if we are going to include it, then lets add test coverage.